storage/client/queries.go

```diff
@@ -394,12 +394,32 @@ func (qf QueryFactory) RuntimeBlocksQuery() string {
 
 func (qf QueryFactory) RuntimeTransactionsQuery() string {
 	return fmt.Sprintf(`
-		SELECT round, tx_index, tx_hash, tx_eth_hash, timestamp, raw, result_raw
+		WITH address_lookups AS (
```

Comment:
Oof, `LIMIT 1000` is high, but regardless, this query is disgusting, even just on `5280995` rows in production.

<details><summary>Query</summary>

```
oasisindexer=> EXPLAIN WITH address_lookups AS (
  SELECT
    txs.round,
    txs.tx_index,
    json_object_agg(pre.address, pre.address_data) AS lookup
  FROM oasis_3.runtime_transactions AS txs
  JOIN oasis_3.runtime_related_transactions AS rel ON txs.round = rel.tx_round
    AND txs.tx_index = rel.tx_index
  JOIN oasis_3.address_preimages AS pre ON rel.account_address = pre.address
  GROUP BY txs.round, txs.tx_index
)
SELECT
    txs.round,
    txs.tx_index,
    txs.tx_hash,
    txs.tx_eth_hash,
    txs.timestamp,
    txs.raw,
    txs.result_raw,
    address_lookups.lookup
  FROM oasis_3.runtime_transactions AS txs
  JOIN address_lookups ON txs.round = address_lookups.round AND txs.tx_index = address_lookups.tx_index
  WHERE (txs.runtime = 'emerald')
ORDER BY txs.runtime DESC, txs.round DESC, txs.tx_index DESC
LIMIT 1000;
```

</details>

<details><summary>Plan</summary>

```                                                                                    QUERY PLAN                                                                                  
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=9659943.41..9667236.21 rows=1000 width=750)
   ->  Merge Join  (cost=9659943.41..11022077.80 rows=186778 width=750)
         Merge Cond: (((txs.round)::bigint = (address_lookups.round)::bigint) AND ((txs.tx_index)::integer = (address_lookups.tx_index)::integer))
         ->  Index Scan Backward using runtime_transactions_pkey on runtime_transactions txs  (cost=0.43..1330276.36 rows=5215771 width=718)
               Index Cond: (runtime = 'emerald'::runtime)
         ->  Sort  (cost=9659942.98..9661246.92 rows=521577 width=44)
               Sort Key: address_lookups.round DESC, address_lookups.tx_index DESC
               ->  Subquery Scan on address_lookups  (cost=2544678.32..9594365.17 rows=521577 width=44)
                     ->  GroupAggregate  (cost=2544678.32..9589149.40 rows=521577 width=44)
                           Group Key: txs_1.round, txs_1.tx_index
                           ->  Merge Join  (cost=2544678.32..8296204.27 rows=171523388 width=82)
                                 Merge Cond: (((txs_1.round)::bigint = (rel.tx_round)::bigint) AND ((txs_1.tx_index)::integer = (rel.tx_index)::integer))
                                 ->  Gather Merge  (cost=626367.93..1233830.41 rows=5215771 width=12)
                                       Workers Planned: 2
                                       ->  Sort  (cost=625367.90..630801.00 rows=2173238 width=12)
                                             Sort Key: txs_1.round, txs_1.tx_index
                                             ->  Parallel Index Only Scan using runtime_transactions_pkey on runtime_transactions txs_1  (cost=0.43..322335.23 rows=2173238 width=12)
                                 ->  Materialize  (cost=1918012.90..4086200.42 rows=18225197 width=82)
                                       ->  Gather Merge  (cost=1918012.90..4040637.43 rows=18225197 width=82)
                                             Workers Planned: 2
                                             ->  Sort  (cost=1917012.88..1935997.46 rows=7593832 width=82)
                                                   Sort Key: rel.tx_round, rel.tx_index
                                                   ->  Hash Join  (cost=1388.08..322420.70 rows=7593832 width=82)
                                                         Hash Cond: ((rel.account_address)::text = (pre.address)::text)
                                                         ->  Parallel Seq Scan on runtime_related_transactions rel  (cost=0.00..301096.32 rows=7593832 width=59)
                                                         ->  Hash  (cost=963.59..963.59 rows=33959 width=70)
                                                               ->  Seq Scan on address_preimages pre  (cost=0.00..963.59 rows=33959 width=70)
(27 rows)
```
</details>

Solutions:

1. The 1:1 or fast query route is possible if we choose not to show the `toEth` address. This wasn't explicitly called out in the ticket, but I think we wanted this, with @pro-wh suggestion.

2. Let me try to subquery my way out of this first?

3. The two queries multiple pass messes up the clean logic of `renderRuntimeTransaction` function, but would be preferable to me than probably more denormalizing.

Reply:
Thank you! Bummer for the results; oh well.

BTW, if you run `EXPLAIN ANALYZE` instead of just `EXPLAIN`, you get actual timing info, which is usually quite helpful. Though here ... there's a step that materializes an estimated 1.8M rows, that's going to be slow no matter what :)

Re 1: I agree `toEth` is very useful, let's try hard to keep it. 
Re 2: Sure. Maybe you can do sth like `WITH txs AS (SELECT FROM oasis_3.runtime_transactions LIMIT 1000) SELECT ... FROM txs JOIN blah JOIN blah`? That will encourage postgres to do index lookups because there's only 1000 rows (x4 JOINs). But don't forget to rebase first; Andy just introduced additional complexity into this query that we'll have to accommodate; I don't remember if it's problematic in that sense. Also, this assumes that the JOINs will not change the number of rows in either direction.
Re 3: We can break `renderRuntimeTransaction` into two stages no problem. What would be even better would be to run most of `renderRuntimeTransaction` at analysis time and modify the table so that it contains a much more structured version of the tx, with each field in a separate column. At render time, we wouldn't need `renderRuntimeTransaction` any more, we'd just pull from the DB directly. Warren initially set up thsi parsing/rendering step outside analysis so we'd have more flexibility in the parsing (eth is tricky, and you don't want to restructure the table every time you figure out how to parse the tx more/better), but we're pretty stable now and it would be cleaner to do the tx parsing during analysis.

Denormalization: I'm not opposed to it philosophically. The indexer uses mostly INSERTs, and only few UPDATEs, so it's unusually easy to keep even fairly denormalized data in sync, I feel.
And FWIW, the last branch `ANALYZE` output builds a denormalized table on the fly: It joins `runtime_related_transactions` and `address_preimages` and materializes the join :). Though actually, doing that denorm ourselves makes it a little trickier later on to implement parallel processing of blocks, because in that scenario, by the time we observe a tx-related address we don't necessarily know its preimage yet. But we could work around it with a final UPDATE after the parallel work is done. So denorm is on the table, even this particular denorm is on the table, it's just not my first choice.

Denormalization 2: The first thing I'd probably recommend going for is getting rid of `runtime_related_transactions` (terrible name btw) and sticking that info into `runtime_transactions` directly as a column, much like [we do for events](https://github.com/oasisprotocol/oasis-indexer/blob/96a140711e77d460da56a0e5f5bf4e6bbaf8283b/storage/migrations/02_oasis_3_runtimes.up.sql#L98-L98). We're using these two separate approaches (`TEXT[]` vs `JOIN`) for random reasons only; it would be nice to settle on one anyway.

If you do any denorm work, or move around the parsing from `renderRuntimeTransaction` (see "re 3" above), that's best done in a separate prework PR.

Reply:
Up to you what you wanted to try first. I mostly wrote a wall of text to nudge us towards changes that will likely be beneficial for this issue but at the same time were on my mental tech-debt wishlist even before the PR.

Also BTW, setting `LIMIT 10` doesn't change the execution plan (or time, which is ~50s). `VACUUM ANALYZE` and `REINDEX SCHEMA` don't help either.

Reply:
> Denormalization 2

This would be my preferred denormalization strategy. I don't see why we couldn't have done that in one pass with `runtime_transactions`.

I may disappoint you though. I think this is looking a lil better. What is our perf expectation 😅?

<details><summary>Query</summary>

```
SELECT 
    txs.round,
    txs.tx_index,
    txs.tx_hash,
    json_object_agg(pre.address, pre.address_data) AS lookup
FROM oasis_3.runtime_transactions AS txs
JOIN oasis_3.runtime_related_transactions AS rel ON txs.round = rel.tx_round
    AND txs.tx_index = rel.tx_index
    AND txs.round = rel.tx_round
JOIN oasis_3.address_preimages AS pre ON rel.account_address = pre.address
WHERE (txs.round, txs.tx_index, txs.tx_hash) IN

(
    SELECT 
    txs.round,
    txs.tx_index,
    txs.tx_hash
    FROM oasis_3.runtime_transactions AS txs
    WHERE (txs.runtime = 'emerald')
    ORDER BY txs.round DESC, txs.tx_index DESC
    LIMIT 1000
)

GROUP BY txs.round, txs.tx_index, txs.tx_hash;
```
</details>

<details><summary>Planner</summary>

```

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=552829.82..552841.47 rows=466 width=109) (actual time=14065.302..14069.309 rows=1000 loops=1)
   Group Key: txs.round, txs.tx_index, txs.tx_hash
   ->  Sort  (cost=552829.82..552830.99 rows=466 width=147) (actual time=14065.181..14065.411 rows=2707 loops=1)
         Sort Key: txs.round, txs.tx_index, txs.tx_hash
         Sort Method: quicksort  Memory: 816kB
         ->  Nested Loop  (cost=8236.11..552809.17 rows=466 width=147) (actual time=3515.101..14061.673 rows=2707 loops=1)
               ->  Hash Join  (cost=8235.69..552339.24 rows=466 width=124) (actual time=3512.093..13829.490 rows=2708 loops=1)
                     Hash Cond: (((rel.tx_round)::bigint = (txs.round)::bigint) AND ((rel.tx_index)::integer = (txs.tx_index)::integer))
                     ->  Seq Scan on runtime_related_transactions rel  (cost=0.00..407409.97 rows=18225197 width=59) (actual time=1.452..8104.875 rows=18220073 loops=1)
                     ->  Hash  (cost=8235.48..8235.48 rows=14 width=89) (actual time=724.715..724.721 rows=1000 loops=1)
                           Buckets: 1024  Batches: 1  Memory Usage: 130kB
                           ->  Nested Loop  (cost=272.98..8235.48 rows=14 width=89) (actual time=5.370..722.444 rows=1000 loops=1)
                                 ->  HashAggregate  (cost=272.98..282.98 rows=1000 width=77) (actual time=2.885..5.002 rows=1000 loops=1)
                                       Group Key: ("ANY_subquery".round)::bigint, ("ANY_subquery".tx_index)::integer, ("ANY_subquery".tx_hash)::text
                                       Batches: 1  Memory Usage: 577kB
                                       ->  Subquery Scan on "ANY_subquery"  (cost=0.43..265.48 rows=1000 width=77) (actual time=1.245..2.157 rows=1000 loops=1)
                                             ->  Limit  (cost=0.43..255.48 rows=1000 width=77) (actual time=1.244..1.880 rows=1000 loops=1)
                                                   ->  Index Scan Backward using runtime_transactions_pkey on runtime_transactions txs_1  (cost=0.43..1330276.36 rows=5215771 width=77) (actual time=1.243..1.765 rows=1000 loops=1)
                                                         Index Cond: (runtime = 'emerald'::runtime)
                                 ->  Index Scan using ix_runtime_transactions_tx_hash on runtime_transactions txs  (cost=0.00..7.94 rows=1 width=77) (actual time=0.713..0.716 rows=1 loops=1000)
                                       Index Cond: ((tx_hash)::text = ("ANY_subquery".tx_hash)::text)
                                       Rows Removed by Index Recheck: 0
                                       Filter: ((("ANY_subquery".round)::bigint = (round)::bigint) AND (("ANY_subquery".tx_index)::integer = (tx_index)::integer))
               ->  Index Scan using address_preimages_pkey on address_preimages pre  (cost=0.41..1.01 rows=1 width=70) (actual time=0.085..0.085 rows=1 loops=2708)
                     Index Cond: ((address)::text = (rel.account_address)::text)
 Planning Time: 13.361 ms
 Execution Time: 14069.796 ms
```
</details>

Reply:
If this isn't fast enough, probably have to denormalize the `runtime_transactions` table to swallow `runtime_related_transactions`.

Reply:
woa, I gotta look into this tomorrow

Reply:
With this new approach, we cannot filter results by related address ([like this](https://github.com/oasisprotocol/oasis-indexer/blob/288d291fca9fc0e8ccb9f5d0ac68fc74932f285b/storage/client/queries.go#L415-L415), I did warn to rebase first somewhere in the wall of text above :)).
It's also still way too slow, unfortunately. I expect this will be one of the most commonly loaded web pages, and so one of the most common queries we run. **Twice**, once we [start returning  the total number of pagination pages](https://app.clickup.com/t/861mcbbhg). So we should be shooting for about 100ms I reckon.

I'm not too worried yet. This task just happened to expose our tech debt and is forcing us to pay it down:
1. Parse txs during analysis time so we can fetch the txs and then join `address_preimages` just to `sender` and `to` fields. (Alternative: Fetch txs into go, parse txs, go to the DB again for preimages. That's uglier though, and I don't think it's easier to implement either.)
2. Put related addresses directly into `runtime_transactions` in a `TEXT[]` column.

Once we have those two PRs done, let's revisit this and see how hard (or not) the problem is.

Reply:
Update: I'm hopeful. Joining 15k addresses with address_preimages takes 40ms. Filtering by a `TEXT[]` field can be mega expensive, but when using a gin index and limiting results to at most 1000 (as we will), the cost is under 20ms. And that should be all we need.

Reply:
> we cannot filter results by related address

I rebased this PR, but didn't include that `WHERE` because I thought that use case was narrow and more performant than the general case. I'd imagine filtering to return the primary key of `(round, tx_index, tx_hash)` works as expected.

> It's also still way too slow, unfortunately. I expect this will be one of the most commonly loaded web pages, and so one of the most common queries we run

Fair. I do think going 1K to 2K for max limit shouldn't change much.

> 1. Parse txs during analysis time so we can fetch the txs and then join address_preimages just to sender and to fields. (Alternative: Fetch txs into go, parse txs, go to the DB again for preimages. That's uglier though, and I don't think it's easier to implement either.)

Slight groan on this haha, but yeah, I do agree we should just pay the one time cost up front.

> 2. Put related addresses directly into runtime_transactions in a TEXT[] column.

Very aligned on this.

Reply:
> I'm quite worried about the performance of this quadruple JOIN.

what was the original quadruple join?

> Staging has 2M runtime txs right now; it will have >10M

maybe this used to be different, but we never have to scan all transactions to complete this query right?

> * stoer related_addresses inside runtime_txs

we do this for events, maybe we can do it here. is it efficient enough to find txs that are related to a given address this way?

> * preimage inside related_addresses

might end up being annoying to update the related addresses table when we find out a new preimage. I hope we don't have to do this :skull:

> * parse the txs to get oasis addresses

don't need to do this in a separate round trip. the accounts related transactions table knows the addresses you'll need

> Materialize  (cost=1918012.90..4086200.42 rows=18225197 width=82)

omg why does it materialize that? and why does it go up from 7593832 to 18225197?

> 1. The 1:1 or fast query route is possible if we choose not to show the `toEth` address. This wasn't explicitly called out in the ticket, but I think we wanted this, with @pro-wh suggestion.

what's the 1:1? what's the fast query route? what's `toEth`? what was @pro-wh suggestion? why does clickup render `inline code` in a proportional font?

> 3\. two queries multiple pass messes up the clean logic

won't need multiple passes because that accounts related transactions table already knows. still two queries though, which should be doable

> there's a step that materializes an estimated 1.8M rows

no 18 million

> by the time we observe a tx-related address we don't necessarily know its preimage yet

that should already be the case. for example, deposits into Emerald are made out to an oasis-style address, and that account may never have been used in an ethereum style transaction before

> random reasons only

being economical in how many postgresql-specific bells and whistles we use :stuck_out_tongue: 

> WHERE (txs.round, txs.tx_index, txs.tx_hash) IN

what's the `WHERE ... IN ...` for? also, take out the `txs.tx_hash`. its making the planner do some extra hashing



> AND txs.round = rel.tx_round

duplicated condition

Reply:
->  Seq Scan on runtime_related_transactions rel  (cost=0.00..407409.97 rows=18225197 width=59) (actual time=1.452..8104.875 rows=18220073 loops=1)

this is where things go wrong btw. we should definitely not be doing a full scan on related transactions. fixing

Reply:
ok please try with #320

Reply:
Post #320 

<details><summary>New</summary>

```
                                                                                                             QUERY PLAN                                                           
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=552740.13..552751.78 rows=466 width=109) (actual time=8004.759..8008.638 rows=1000 loops=1)
   Group Key: txs.round, txs.tx_index, txs.tx_hash
   ->  Sort  (cost=552740.13..552741.30 rows=466 width=147) (actual time=8004.693..8004.912 rows=2707 loops=1)
         Sort Key: txs.round, txs.tx_index, txs.tx_hash
         Sort Method: quicksort  Memory: 816kB
         ->  Nested Loop  (cost=8236.11..552719.48 rows=466 width=147) (actual time=2774.860..8000.603 rows=2707 loops=1)
               ->  Hash Join  (cost=8235.69..552249.55 rows=466 width=124) (actual time=2771.679..7761.576 rows=2708 loops=1)
                     Hash Cond: (((rel.tx_round)::bigint = (txs.round)::bigint) AND ((rel.tx_index)::integer = (txs.tx_index)::integer))
                     ->  Seq Scan on runtime_related_transactions rel  (cost=0.00..407358.72 rows=18220072 width=59) (actual time=0.104..3621.805 rows=18220073 loops=1)
                     ->  Hash  (cost=8235.48..8235.48 rows=14 width=89) (actual time=721.130..721.134 rows=1000 loops=1)
                           Buckets: 1024  Batches: 1  Memory Usage: 130kB
                           ->  Nested Loop  (cost=272.98..8235.48 rows=14 width=89) (actual time=11.787..718.852 rows=1000 loops=1)
                                 ->  HashAggregate  (cost=272.98..282.98 rows=1000 width=77) (actual time=4.259..5.872 rows=1000 loops=1)
                                       Group Key: ("ANY_subquery".round)::bigint, ("ANY_subquery".tx_index)::integer, ("ANY_subquery".tx_hash)::text
                                       Batches: 1  Memory Usage: 577kB
                                       ->  Subquery Scan on "ANY_subquery"  (cost=0.43..265.48 rows=1000 width=77) (actual time=1.447..2.856 rows=1000 loops=1)
                                             ->  Limit  (cost=0.43..255.48 rows=1000 width=77) (actual time=1.445..2.427 rows=1000 loops=1)
                                                   ->  Index Scan Backward using runtime_transactions_pkey on runtime_transactions txs_1  (cost=0.43..1330276.36 rows=5215771 width=77) (actual time=1.444..2.258 rows=1000 loops=1)
                                                         Index Cond: (runtime = 'emerald'::runtime)
                                 ->  Index Scan using ix_runtime_transactions_tx_hash on runtime_transactions txs  (cost=0.00..7.94 rows=1 width=77) (actual time=0.708..0.711 rows=1 loops=1000)
                                       Index Cond: ((tx_hash)::text = ("ANY_subquery".tx_hash)::text)
                                       Rows Removed by Index Recheck: 0
                                       Filter: ((("ANY_subquery".round)::bigint = (round)::bigint) AND (("ANY_subquery".tx_index)::integer = (tx_index)::integer))
               ->  Index Scan using address_preimages_pkey on address_preimages pre  (cost=0.41..1.01 rows=1 width=70) (actual time=0.087..0.087 rows=1 loops=2708)
                     Index Cond: ((address)::text = (rel.account_address)::text)
 Planning Time: 16.883 ms
 Execution Time: 8009.534 ms
(27 rows)
```
</details>

Much faster, but still meh.

Reply:
Blarf! Sorry, sorry, I had a dup in my query.

<details><summary>Revised query</summary>

```
SELECT 
    txs.round,
    txs.tx_index,
    txs.tx_hash,
    json_object_agg(pre.address, pre.address_data) AS lookup
FROM oasis_3.runtime_transactions AS txs
JOIN oasis_3.runtime_related_transactions AS rel
    ON txs.runtime = rel.runtime
    AND txs.tx_index = rel.tx_index
    AND txs.round = rel.tx_round
JOIN oasis_3.address_preimages AS pre ON rel.account_address = pre.address
WHERE (txs.round, txs.tx_index, txs.tx_hash) IN

(
    SELECT 
    txs.round,
    txs.tx_index,
    txs.tx_hash
    FROM oasis_3.runtime_transactions AS txs
    WHERE (txs.runtime = 'emerald')
    ORDER BY txs.round DESC, txs.tx_index DESC
    LIMIT 1000
)

GROUP BY txs.round, txs.tx_index, txs.tx_hash;
```
</details>


<details><summary>Res</summary>

```
                                                                                                          QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=8314.13..8315.36 rows=49 width=109) (actual time=45.768..49.623 rows=1000 loops=1)
   Group Key: txs.round, txs.tx_index, txs.tx_hash
   ->  Sort  (cost=8314.13..8314.26 rows=49 width=147) (actual time=45.745..45.980 rows=2707 loops=1)
         Sort Key: txs.round, txs.tx_index, txs.tx_hash
         Sort Method: quicksort  Memory: 816kB
         ->  Nested Loop  (cost=273.96..8312.76 rows=49 width=147) (actual time=2.133..42.887 rows=2707 loops=1)
               ->  Nested Loop  (cost=273.54..8263.34 rows=49 width=124) (actual time=2.077..14.216 rows=2708 loops=1)
                     ->  Nested Loop  (cost=272.98..8235.48 rows=14 width=93) (actual time=1.629..7.359 rows=1000 loops=1)
                           ->  HashAggregate  (cost=272.98..282.98 rows=1000 width=77) (actual time=1.613..2.151 rows=1000 loops=1)
                                 Group Key: ("ANY_subquery".round)::bigint, ("ANY_subquery".tx_index)::integer, ("ANY_subquery".tx_hash)::text
                                 Batches: 1  Memory Usage: 577kB
                                 ->  Subquery Scan on "ANY_subquery"  (cost=0.43..265.48 rows=1000 width=77) (actual time=0.024..0.997 rows=1000 loops=1)
                                       ->  Limit  (cost=0.43..255.48 rows=1000 width=77) (actual time=0.023..0.658 rows=1000 loops=1)
                                             ->  Index Scan Backward using runtime_transactions_pkey on runtime_transactions txs_1  (cost=0.43..1330276.36 rows=5215771 width=77) (actual time=0.023..0.544 rows=1000 loops=1)
                                                   Index Cond: (runtime = 'emerald'::runtime)
                           ->  Index Scan using ix_runtime_transactions_tx_hash on runtime_transactions txs  (cost=0.00..7.94 rows=1 width=81) (actual time=0.004..0.005 rows=1 loops=1000)
                                 Index Cond: ((tx_hash)::text = ("ANY_subquery".tx_hash)::text)
                                 Rows Removed by Index Recheck: 0
                                 Filter: ((("ANY_subquery".tx_index)::integer = (tx_index)::integer) AND (("ANY_subquery".round)::bigint = (round)::bigint))
                     ->  Index Scan using ix_runtime_related_transactions_round_index on runtime_related_transactions rel  (cost=0.56..1.97 rows=2 width=63) (actual time=0.005..0.006 rows=3 loops=1000)
                           Index Cond: ((runtime = txs.runtime) AND ((tx_round)::bigint = (txs.round)::bigint) AND ((tx_index)::integer = (txs.tx_index)::integer))
               ->  Index Scan using address_preimages_pkey on address_preimages pre  (cost=0.41..1.01 rows=1 width=70) (actual time=0.010..0.010 rows=1 loops=2708)
                     Index Cond: ((address)::text = (rel.account_address)::text)
 Planning Time: 1.591 ms
 Execution Time: 49.836 ms
(25 rows)
```
</details>

Reply:
> > I'm quite worried about the performance of this quadruple JOIN.
> 
> what was the original quadruple join?

I think he meant the `JOIN` across `runtime_transactions`, `runtime_related_transactions`, `address_preimages`, and on the subquery again.
> 
> > Staging has 2M runtime txs right now; it will have >10M
> 
> maybe this used to be different, but we never have to scan all transactions to complete this query right?

I believe this.
> 
> > * stoer related_addresses inside runtime_txs
> 
> we do this for events, maybe we can do it here. is it efficient enough to find txs that are related to a given address this way?

addresses could be one to many right? I think reaching across a different table is okay.

> 
> > * preimage inside related_addresses
> 
> might end up being annoying to update the related addresses table when we find out a new preimage. I hope we don't have to do this 💀

Agree.
> 
> > * parse the txs to get oasis addresses
> 
> don't need to do this in a separate round trip. the accounts related transactions table knows the addresses you'll need

Aye.
> 
> > Materialize  (cost=1918012.90..4086200.42 rows=18225197 width=82)
> 
> omg why does it materialize that? and why does it go up from 7593832 to 18225197?

Bad querying from moi.
> 
> > 1. The 1:1 or fast query route is possible if we choose not to show the `toEth` address. This wasn't explicitly called out in the ticket, but I think we wanted this, with @pro-wh suggestion.
> 
> what's the 1:1? what's the fast query route? what's `toEth`? what was @pro-wh suggestion? why does clickup render `inline code` in a proportional font?

1:1 was referring to the `JOIN` to the `runtime_transaction_signers` to get the address in the first version of this PR, but @pro-wh helpfully called out that the `toEth` address is meaningful. "why does clickup _______?"
> 
> > 3. two queries multiple pass messes up the clean logic
> 
> won't need multiple passes because that accounts related transactions table already knows. still two queries though, which should be doable

K.
> 
> > there's a step that materializes an estimated 1.8M rows
> 
> no 18 million

Self-cringe. It is better now.
> 
> > by the time we observe a tx-related address we don't necessarily know its preimage yet
> 
> that should already be the case. for example, deposits into Emerald are made out to an oasis-style address, and that account may never have been used in an ethereum style transaction before
> 
> > random reasons only
> 
> being economical in how many postgresql-specific bells and whistles we use 😛
> 
> > WHERE (txs.round, txs.tx_index, txs.tx_hash) IN
> 
> what's the `WHERE ... IN ...` for? also, take out the `txs.tx_hash`. its making the planner do some extra hashing

K. I should just use the correct primary key.
> 
> > AND txs.round = rel.tx_round
> 
> duplicated condition

Yes.

Reply:
:raised_hands: thanks!

Reply:
Thank you Warren for realizing the bad query plan was not because postgres was too stupid to use index, or because the indexes were too expensive, but because they were missing :facepalm: 

I'm a little sad for losing the opportunity/excuse to do the cleanup we discussed :), but it's good news that we can move ahead faster.

Reply:
Not anti-cleanup, but mounting guilt from not finishing tickets, maybe in a few sprints 🙏